### PR TITLE
Batch email change category

### PR DESCRIPTION
**Pull request by mchojrin**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 

## Description

This PR addresses the feature discussed [here](https://forum.mautic.org/t/batch-email-category-update/33106/6):

To allow changing the email category for multiple messages at a time.

The UI looks like this:

image
image

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Login
3. Go to /s/emails
4. If you don't have any email written create a few and put them in different categories
5. From the email list, select some
6. Click on the batch actions menu
7. Select *Change Category*
8. Select a category
9. Save
10. Refresh the page
11. Every previously selected email should belong to the selected category

### CODE REVIEW

**Comment by andersonjeccel** [comment]
@mchojrin Hey, is this ready for testing?

**Comment by mchojrin** [comment]
@andersonjeccel I think so... I run the tests and they're all passing. Is there any other requirement prior to testing?

**Comment by andersonjeccel** [comment]
Got it!

The main step to have other devs reviewing is to properly fill the description

yeah, it’s boring but contains the minimum basic they need to know about the code (I’ve been creating ~175 PRs for Mautic in the last 6 months, believe me that it’s needed)

would you be up for that?
(then I can assign some people to review and the right labels)

**Comment by mchojrin** [comment]
@andersonjeccel sure, no problem. I'll get to it ;)

**Comment by mchojrin** [comment]
There, @andersonjeccel is this good?

**Comment by andersonjeccel** [comment]
@mchojrin Perfect, thanks!

---

I tested the PR here and got an error, have a look:

https://github.com/user-attachments/assets/4860be2b-a47f-4cef-b527-7f30735b6238

Steps I tried:
1. Go to Emails, create new and select the theme Brienz
2. Fill subject and name using `Email 16/07/2024, 08:40:24` then Save & Close
3. Go to Categories, create new and go back to Emails
4. Select using the Select all checkbox, open the dropdown and click Change category
5. Select the new category and click Save
6. The message will appear

Do you have any idea why?

**Comment by mchojrin** [comment]
@andersonjeccel I just tried the exact steps you mentioned and it worked alright... Maybe you need to refresh the cache?

**Comment by andersonjeccel** [comment]
@mchojrin Hm, tried both prod and dev environments...
After clearing the cache by manually deleting the cache folder and running `ddev exec bin/console cache:clear`, it starts showing this way:

image

It's a clean DDEV install (including database)

Then tried on Gitpod:

image

image

What if you try to clear your cache? Perhaps the issue might appear

**Comment by mchojrin** [comment]
@andersonjeccel wow.. ok.. I'll try to test in different environments and get back here.

Thanks!

**Comment by andersonjeccel** [comment]
I hope we can solve, there was a lot of your work on it :)
call me if you need something

**Comment by mchojrin** [comment]
Hey @andersonjeccel just tried again clearing the local cache and on a fresh new gitpod and couldn't reproduce.

[Here](https://www.loom.com/share/cf9d70d778f84d1992508b6e17f135ee?sid=fb443825-4e39-4a73-9240-48d6ec449448)'s my video

I did not an unexpected behaviour that I'll address though, the select category modal is showing all available categories instead of only email ones.

I'll leave the gitpod up if you want to try it https://8080-mchojrin-mautic-b5nqmcv3vbe.ws-eu115.gitpod.io/s/emails

**Comment by mchojrin** [comment]
I just checked my code and it is applying the filter by email bundle to the categories shown... perhaps it's a problem with `\Mautic\CategoryBundle\Entity\CategoryRepository::getCategoryList`?

**Comment by andersonjeccel** [comment]
@escopecz @kuzmany Would you like to have a look?

File: .ddev/docker-compose.selenium-chrome.yaml [hunk 1]
```diff
@@ -15,6 +15,7 @@ services:
       - VIRTUAL_HOST=$DDEV_HOSTNAME
       - HTTPS_EXPOSE=7900:7900
       - HTTP_EXPOSE=7910:7900
+      - VNC_NO_PASSWORD=1
```
Comment (escopecz) [inline .ddev/docker-compose.selenium-chrome.yaml]: How is this related to this PR?

File: app/bundles/EmailBundle/Assets/js/batch.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+            }
+
+            return true;
+        }
+
+    }
+
+    mQuery('#MauticSharedModal').modal('hide');
+
+    return false;
```
Comment (escopecz) [inline app/bundles/EmailBundle/Assets/js/batch.js:15]: Are these return statements used for anything?

File: app/bundles/EmailBundle/Assets/js/batch.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+//EmailBundle (Copied from app/bundles/LeadBundle/Assets/js/lead.js)
+Mautic.emailBatchSubmit = function() {
+    if (Mautic.batchActionPrecheck()) {
+        if (mQuery('#email_batch_newCategory').val()) {
+            if (mQuery('#email_batch_ids').length) {
+                mQuery('#email_batch_ids').val(Mautic.getCheckedListIds(false, true));
```
Comment (escopecz) [inline app/bundles/EmailBundle/Assets/js/batch.js]: Selecting HTML elements via jQuery is an expensive operation. Use variables instead.
```suggestion
            const $emailBatchIds = mQuery('#email_batch_ids');
            if ($emailBatchIds.length) {
                $emailBatchIds.val(Mautic.getCheckedListIds(false, true));
```
The `$` is a naming convention to distinguish jQuery objects.

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+
+class BatchEmailController extends AbstractFormController
+{
+    public function __construct(
+        private readonly EmailActionModel $actionModel,
+        private readonly CategoryModel $categoryModel,
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: The `readonly` properties are supported since PHP 8.1. We'd have to merge https://github.com/mautic/mautic/pull/13972 first to keep this.

File: app/bundles/EmailBundle/Model/EmailActionModel.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Mautic\EmailBundle\Model;
+
+use Mautic\CategoryBundle\Entity\Category;
+
+readonly class EmailActionModel
```
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailActionModel.php]: `readonly` classes are supported since PHP 8.2. Mautic supports PHP 8.0 at the moment. See the composer.json.

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php:2]: Please add `declare(strict_types=1);` to all new PHP classes.
```suggestion

declare(strict_types=1);

```

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+     * Adds or removes categories to multiple emails defined by email ID.
+     */
+    public function execAction(Request $request): JsonResponse
+    {
+        $params = $request->get('email_batch');
+        $ids    = empty($params['ids']) ? [] : json_decode($params['ids']);
+
+        if ($ids && is_array($ids)) {
+            $newCategoryId = $params['newCategory'];
+            $emailIds      = json_decode($params['ids']);
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: How is `$ids` different from `$emailIds`?

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+            $affected = $this->actionModel->setCategory($emailIds, $this->categoryModel->getEntity($newCategoryId));
+
+            $newCategory = current($affected)?->getCategory();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: Why getting the new category entity this way? You are loading it on line 55. Why not store it to a variable when it's being loaded there?

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            'flashes'     => $this->getFlashContent(),
+            'affected'    => !empty($affected) ? array_map(fn (Email $affected) => $affected->getId(), $affected) : [],
+            'newCategory' => !empty($newCategory) ? $newCategory->getTitle() : null,
+        ]);
+    }
+
+    /**
+     * View the modal form for adding contacts into categories in batches.
+     *
+     * @todo Inject the selected email ids into the view
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: Please resolve all todos

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+use Mautic\CoreBundle\Controller\AbstractFormController;
+use Mautic\CoreBundle\Factory\MauticFactory;
+use Mautic\CoreBundle\Factory\ModelFactory;
+use Mautic\CoreBundle\Helper\CoreParametersHelper;
+use Mautic\CoreBundle\Helper\UserHelper;
+use Mautic\CoreBundle\Security\Permissions\CorePermissions;
+use Mautic\CoreBundle\Service\FlashBag;
+use Mautic\CoreBundle\Translation\Translator;
+use Mautic\EmailBundle\Entity\Email;
+use Mautic\EmailBundle\Form\Type\BatchCategoryType as BatchType;
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: Why the alias? It's hard to read later on.

File: app/bundles/EmailBundle/Controller/BatchEmailController.php [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            ]);
+        } else {
+            $this->addFlashMessage('mautic.core.error.ids.missing');
+        }
+
+        return new JsonResponse([
+            'closeModal'  => true,
+            'flashes'     => $this->getFlashContent(),
+            'affected'    => !empty($affected) ? array_map(fn (Email $affected) => $affected->getId(), $affected) : [],
+            'newCategory' => !empty($newCategory) ? $newCategory->getTitle() : null,
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/BatchEmailController.php]: I can't find where are these 2 params used. Can those be deleted?

File: app/bundles/EmailBundle/Controller/EmailController.php [hunk 12]
```diff
@@ -1405,6 +1405,8 @@ public function sendAction(Request $request, $objectId)
      * Deletes a group of entities.
      *
      * @return Response
+     *
+     * @todo Move this to BatchEmailController
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/EmailController.php]: Please resolve this todo

File: app/bundles/EmailBundle/Form/Type/BatchCategoryType.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            [
+                'items',
+            ]
+        );
+    }
+
+    /**
+     * @return string
+     */
+    public function getBlockPrefix()
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/BatchCategoryType.php]:
```suggestion
    public function getBlockPrefix(): string
```

File: app/bundles/EmailBundle/Form/Type/BatchCategoryType.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $builder->add(
+            'newCategory',
+            ChoiceType::class,
+            [
+                'label'             => 'mautic.email.category.batch.set',
+                'multiple'          => false,
+                'choices'           => $options['items'],
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/BatchCategoryType.php]: Is there any reason why the list of categories is passed as an option and not loaded from the database in this form class?

File: app/bundles/EmailBundle/Model/EmailActionModel.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public function __construct(
+        private EmailModel $emailModel
+    ) {
+    }
+
+    public function setCategory(array $emailsIds, Category $newCategory): array
+    {
+        $emails = $this->emailModel->getByIds($emailsIds);
```
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailActionModel.php]:
```suggestion
        $emails = $this->emailModel->getRepository->findBy(['id' => $emailsIds]);
```
Then the whole `getByIds()` method can be deleted.

File: app/bundles/EmailBundle/Model/EmailActionModel.php [hunk 16]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Mautic\EmailBundle\Model;
+
+use Mautic\CategoryBundle\Entity\Category;
+
+readonly class EmailActionModel
+{
+    public function __construct(
+        private EmailModel $emailModel
```
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailActionModel.php]: Ideally, this class should require the `EmailModel` (to save entities), `EmailRepository` (to load entities) and `CorePermissions` (to check permissions) to do what's necessary. Adding more methods to the bloated `EmailModel` is not a good idea. Especially when you have this nice new model.

File: app/bundles/EmailBundle/Resources/views/Batch/form.html.twig [hunk 17]
```diff
@@ -0,0 +1,7 @@
+{#
+  Variables
+    - form
+#}
+
+{% set attr = form.vars.attr|merge({'data-submit-callback': 'emailBatchSubmit'}) %}
```
Comment (escopecz) [inline app/bundles/EmailBundle/Resources/views/Batch/form.html.twig]: Why not set the `data-submit-callback` param in the form type itself?

File: app/bundles/EmailBundle/Resources/views/Email/select_category.html.twig [hunk 18]
```diff
@@ -0,0 +1 @@
+{{form(form)}}
```
Comment (escopecz) [inline app/bundles/EmailBundle/Resources/views/Email/select_category.html.twig]: Where is this template used?

File: app/bundles/EmailBundle/Tests/Controller/EmailControllerTest.php [hunk 19]
```diff
@@ -30,12 +31,10 @@
...
 
-class EmailControllerTest extends \PHPUnit\Framework\TestCase
+class EmailControllerTest extends TestCase
 {
-    /**
-     * @var MockObject|Translator
-     */
-    private MockObject $translatorMock;
+    public const NEW_CATEGORY_TITLE = 'New category';
+    private MockObject|Translator $translatorMock;
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Controller/EmailControllerTest.php]: It is actually a type of `MockObject&Translator` which isn't possible to do in PHP 8.0. Please revert it back to a comment type. Same for row 47

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class EmailActionModelTest extends TestCase
+{
+    public const NEW_CATEGORY_TITLE = 'New category';
+    public const OLD_CATEGORY_TITLE = 'Old category';
+
+    /**
+     * @var MockObject|EmailModel
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php]:
```suggestion
     * @var MockObject&EmailModel
```

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        parent::setUp();
+
+        $this->emailModel       = $this->createMock(EmailModel::class);
+        $this->emailActionModel = new EmailActionModel($this->emailModel);
+    }
+
+    /**
+     * @test
+     */
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php]: Is this comment adding any value?
```suggestion
```
Applies to other places

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+namespace Mautic\EmailBundle\Tests\Model;
+
+use Mautic\CategoryBundle\Entity\Category;
+use Mautic\EmailBundle\Entity\Email;
+use Mautic\EmailBundle\Model\EmailActionModel;
+use Mautic\EmailBundle\Model\EmailModel;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class EmailActionModelTest extends TestCase
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: It is better to write functional tests as those will ensure the new feature works as it makes real DB queries and test the input/output. In unit tests there is a lot of assumptions. And in many cases, functional tests are faster to write as there is no need to write mocks. But I can see you are also adding e2e tests, so that would replace functional tests.

File: tests/_support/Page/Acceptance/EmailsPage.php [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+     * and use it in tests like: Page\Edit::route('/123-post');.
+     */
+    public static function route($param)
+    {
+        return static::$URL.$param;
+    }
+
+    /**
+     * @var AcceptanceTester;
+     */
```
Comment (escopecz) [inline tests/_support/Page/Acceptance/EmailsPage.php]:
```suggestion
```

File: tests/acceptance/EmailManagementCest.php [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+        $I->click(EmailsPage::$selectedActionsDropdown);
+
+        $I->waitForElementClickable(EmailsPage::$changeCategoryAction);
+        $I->click(EmailsPage::$changeCategoryAction);
+
+        $I->waitForElementClickable(EmailsPage::$newCategoryDropdown);
+
+        // Click the dropdown menu
+        $I->click('#email_batch_newCategory_chosen > a > span');
+        $I->waitForElementVisible('#email_batch_newCategory_chosen > div > ul > li.active-result');
```
Comment (escopecz) [inline tests/acceptance/EmailManagementCest.php]: Shouldn't the test assert that the email has the category in the end?

**Review by escopecz — CHANGES_REQUESTED** [review]
What a nice feature! I think it can be simplified a bit. Be careful about what PHP versions this project supports as you are a bit in the future ;)

Let me know if my comments make sense.

File: .ddev/docker-compose.selenium-chrome.yaml (hunk 1 above)
Comment (mchojrin) [reply .ddev/docker-compose.selenium-chrome.yaml]: Oh... you're right... it's not :stuck_out_tongue: it was just something I needed to test locally and didn't realize it was part of the PR. Sorry!

File: app/bundles/EmailBundle/Assets/js/batch.js (hunk 2 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Assets/js/batch.js:15]: mmm, not sure. This part I copied from a similar screen

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 4 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: Oh, I assumed the php version was correct since I used the interpreter from docker.

File: app/bundles/EmailBundle/Model/EmailActionModel.php (hunk 5 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Model/EmailActionModel.php]: Got it. It's doesn't _have to_ be readonly anyway

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 8 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: Interesting... I didn't see that but yeah, makes sense

Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: I just remembered, the intention was to test whether there was any email affected, assuming the `setCategory` would only return those where something was changed. Anyway, the category doesn't need to be calculated like this so I'll change it.

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 7 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: They don't. I'll remove it

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 10 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: No problem, I'll remove it

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 11 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: The idea of having these is to allow a future improvement where the UI can be refreshed upon receiving the response from Ajax, without manually refreshing the page.

File: app/bundles/EmailBundle/Model/EmailActionModel.php (hunk 16 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Model/EmailActionModel.php]: I'll see about this

File: app/bundles/EmailBundle/Resources/views/Batch/form.html.twig (hunk 17 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Resources/views/Batch/form.html.twig]: Could be... I guess this is the result of copy & paste from an existing template

File: app/bundles/EmailBundle/Assets/js/batch.js (hunk 2 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Assets/js/batch.js:15]: I get that from the comment. Some old parts of Mautic aren't really up to the standard we want the new code to be in. It's always good to take a look if it can be improved after pasting. As some point someone will copy this code from you and it would be nice to have it in the best shape possible :)

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 4 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: We use the latest PHP version in docker, but still someone may run it on some older PHP version. We want to support all [active PHP versions](https://www.php.net/supported-versions.php). That's why I want to drop support for PHP 8.0 so we could start using the PHP 8.1 syntax. Just have to finish up that PR.

File: tests/acceptance/EmailManagementCest.php (hunk 24 above)
Comment (mchojrin) [reply tests/acceptance/EmailManagementCest.php]: Yes... I guess after having so many fights trying to get the click to work I forgot about the reason for the test.

Thanks, I'll fix it

File: app/bundles/EmailBundle/Assets/js/batch.js (hunk 2 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Assets/js/batch.js:15]: I'll keep that in mind for upcoming PRs ;)

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 11 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: Let's add it then. There is a very small chance that someone will get to implement it afterwards and in the mean time the code could be simpler without it.

File: app/bundles/EmailBundle/Form/Type/BatchCategoryType.php [hunk 25]
```diff
@@ -0,0 +1,65 @@
+<?php
+
```
Comment (shinde-rahul) [inline app/bundles/EmailBundle/Form/Type/BatchCategoryType.php:2]: Please add `declare(strict_types=1);`

File: app/bundles/EmailBundle/Model/EmailActionModel.php [hunk 26]
```diff
@@ -0,0 +1,35 @@
+<?php
+
```
Comment (shinde-rahul) [inline app/bundles/EmailBundle/Model/EmailActionModel.php:2]: Please add `declare(strict_types=1);`

**Review by shinde-rahul — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: I usually prefer to write both but I see your point. If the preferred way is to go with only e2e I can get onboard with that ;)

Comment (escopecz) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Over the years I've learned that if you have a bunch of unit tests covering a feature then it doesn't mean the feature is working. If you refactor the feature, you have to refactor also the unit tests because they contain a lot of implementation details. So it adds work and it won't catch all the bugs.

On the other hand, if you have a functional test that is sending a specific input and asserting for specific output or database state then you are 99% sure that the feature works. And when you refactor the feature, you don't have to touch the tests and be sure that the feature still works.

Comment (mchojrin) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: No problem. Do you think it'd be better to remove this class altogether?

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 11 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: Ok... I'll need to dig into the templates to see how this could be implemented but ok

File: app/bundles/EmailBundle/Form/Type/BatchCategoryType.php (hunk 14 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Form/Type/BatchCategoryType.php]: mmm, probably another copy & paste

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: If it will be covered by a functional test for both scenarios where the user have the permission and doesn't have it then this unit test can be removed.

**Comment by andersonjeccel** [comment]
@escopecz @shinde-rahul nice to see you checking here! thanks for your time

is the PR test working for you? even after those changes, I still get the same  `500 Internal Server Error - PHP Warning - Undefined property: Mautic\EmailBundle\Model\EmailModel::$getRepository` 

- Tried to run composer install again
- Cleared cache manually
- Removed var/cache completely

All with no success :/
I'm using PHP Version 8.3.10

File: app/bundles/EmailBundle/Model/EmailActionModel.php [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+readonly class EmailActionModel
+{
+    public function __construct(
+        private EmailModel $emailModel
+    ) {
+    }
+
+    public function setCategory(array $emailsIds, Category $newCategory): array
+    {
+        $emails = $this->emailModel->getRepository->findBy(['id' => $emailsIds]);
```
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailActionModel.php]:
```suggestion
        $emails = $this->emailModel->getRepository()->findBy(['id' => $emailsIds]);
```
@andersonjeccel my bad. I think that this PR is still work in progress. Wait for the CI checks to be green to retest.

Comment (andersonjeccel) [reply app/bundles/EmailBundle/Model/EmailActionModel.php]: Got it!
Thanks for replying

File: app/bundles/EmailBundle/Assets/js/batch.js (hunk 2 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Assets/js/batch.js:15]: From what I can see, the return value and the hiding of the modal would be needed if there was no email selected. Something that shouldn't happen in the first place as the batch actions wouldn't have been shown.

Still, since the method `emailBatchSubmit` is invoked upon submitting the form to determine whether to submit or not, I don't think it's a bad idea to leave the `return false` in place.

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: I'm not sure about this. I usually try to keep my e2e tests to cover the happy path and deal with edge cases at a lower level. Mainly to prevent e2e suites to be too heavy/time consuming to run.

File: app/bundles/EmailBundle/Controller/EmailController.php (hunk 12 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/EmailController.php]: This action was there before me. I just thought it'd be a good idea to move it to a more proper place but, if we want to be strict, that has nothing to do with this feature

File: app/bundles/EmailBundle/Resources/views/Email/select_category.html.twig (hunk 18 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Resources/views/Email/select_category.html.twig]: Apparently never :stuck_out_tongue: 

I'll remove it

File: app/bundles/EmailBundle/Controller/BatchEmailController.php (hunk 11 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/BatchEmailController.php]: Done!

File: app/bundles/EmailBundle/Form/Type/BatchCategoryType.php (hunk 14 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Form/Type/BatchCategoryType.php]: Solved

File: app/bundles/EmailBundle/Resources/views/Batch/form.html.twig (hunk 17 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Resources/views/Batch/form.html.twig]: Fixed

**Comment by mchojrin** [comment]
Ok, I think I fixed every point if anyone else wants to take a look

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Then a functional test in PHPUNIT would be the way to go.

File: app/bundles/EmailBundle/Controller/EmailController.php (hunk 12 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Controller/EmailController.php]: I get that, but no one will do this in the future if you won't 🙂 So I'm wondering if this comment adds any value.

Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/EmailController.php]: I see... well, I guess for the time being I'll remove it so I don't go out of scope.

Comment (mchojrin) [reply app/bundles/EmailBundle/Controller/EmailController.php]: Done. Though the fact that I removed lines *1408* and *1409* kind of hints that there might be too many things here, doesn't it? :stuck_out_tongue:

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: I'll see about that. Thanks

File: app/bundles/EmailBundle/Controller/EmailController.php (hunk 12 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Controller/EmailController.php]: 100%! A refactoring is necessary.

**Comment by andersonjeccel** [comment]
image

It's working now, showing only email-specific categories for selection and the change is visible immediately after applying!

@mchojrin Tell me when the PR is finished (probably needs only the PHPUnit test to finish as John said)

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (mchojrin) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Hey @escopecz I just checked the code and I noticed there was already a test `\Mautic\EmailBundle\Tests\Model\EmailActionModelTest::testDoesntSetNewCategoryForNonEditableEmails`, would this do?

Comment (escopecz) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Can you link where it is? I can't find it

Comment (andersonjeccel) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: @escopecz 
https://github.com/mchojrin/mautic/blob/7373d582612ba56810423eb5772032aba24154f1/app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php

Comment (andersonjeccel) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Running a composer test locally, there are still some tests failing

Comment (andersonjeccel) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: @mchojrin Hey, got some minutes to have a look?

**Comment by andersonjeccel** [comment]
@escopecz Could you approve the workflow? We need to know if the PR is passing all checks

I think the author is almost giving up bc of this delay

File: app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php (hunk 22 above)
Comment (andersonjeccel) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: @mchojrin Please don't give up, we're almost there! 🙏

Comment (escopecz) [reply app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php:16]: Apologies for late replies. I got a bit distracted with work. I have several PRs on my todo list to re-review. I think I'll find a few hours today.

**Review by escopecz — DISMISSED** [review]
I've simplified the code a bit more and also added an option to assign not only categories type of email but also type of global as the global type should be able to contain all the entities.

I also removed the placeholder in the category select box as that was showing as another option that could be saved but that didn't work. Users must select one category.

@mchojrin could you please check the failing `EmailManagementCest` e2e test? Once that is passing this is good to go.

**Comment by mchojrin** [comment]
Hey @escopecz I had some issues with ddev, I think I got it all sorted out now so hopefully I'll be able to fix the test soon

**Comment by mchojrin** [comment]
I fixed the E2E test, it broke because of changes to the UI... I tried to use element ids as much as possible to avoid this problem but I'm afraid I'll still happen for those elements that don't have one (Such as the batch action menu for emails).

Github is reporting a conflict in the `app/bundles/LeadBundle/Translations/en_US/messages.ini` that I'm not sure how to solve...

**Comment by escopecz** [comment]
I resolved the conflict by keeping the translation for both branches.

**Comment by escopecz** [comment]
BTW, if there is an ID or a class missing on an HTML element, feel free to add it. It helps not only to have more stable tests but also the UI team as they can target that HTML element.

**Comment by mchojrin** [comment]
> BTW, if there is an ID or a class missing on an HTML element, feel free to add it. It helps not only to have more stable tests but also the UI team as they can target that HTML element.

Makes sense :)

I guess I'll keep that in mind moving forward so this PR can finally be integrated.

Thanks!

**Review by escopecz — APPROVED** [review]
This is all good from the code perspective. I also tested the feature for the last time and didn't find any issue.
Screenshot 2024-09-19 at 12 58 17

It degrades test coverage slightly but it does have e2e tests attached which do cover the code. So this time I'm OK with the metric going down as I know the code is actually covered. But it would be great to make functional tests next time too.

Thank you! 👍

**Review by andersonjeccel — APPROVED** [review]
Works perfectly, thanks for the improvement @mchojrin!

image

**Comment by RCheesley** [comment]
@all-contributors please add @mchojrin for code.

Thanks so much for the contribution @mchojrin and welcome to the community!

In the future it'd be nice to add this kind of behaviour to our user docs but at present we don't have anything about working with list view even, so we're good to merge without it for now.

**Comment by escopecz** [comment]
@mchojrin could you please validate a bug that these changes might have introduced? https://github.com/mautic/mautic/issues/14323

File: app/bundles/EmailBundle/Resources/views/Email/_list.html.twig [hunk 28]
```diff
@@ -7,8 +7,24 @@
...
+                                'data-target': '#MauticSharedModal',
+                                'href': path('mautic_email_batch_categories_view'),
+                                'data-header': 'mautic.email.batch.categories'|trans,
+                            },
+                            'btnText': 'mautic.email.batch.categories'|trans,
+                            'iconClass': 'ri-folder-line',
+                        },
+                    ],
+                    'langVar': 'lead.lead',
+                    'routeBase': 'contact',
```
Comment (andersonjeccel) [inline app/bundles/EmailBundle/Resources/views/Email/_list.html.twig:26]: @escopecz this seems to be the specific line that makes the Delete button delete contacts instead of emails
